managed/services/management/dbaas/dbaas_initializer.go

```diff
@@ -0,0 +1,251 @@
+// Copyright (C) 2017 Percona LLC
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+package dbaas
+
+import (
+	"context"
+	"fmt"
+	"math/rand"
+	"sync"
+	"time"
+
+	goversion "github.com/hashicorp/go-version"
+	dbaascontrollerv1beta1 "github.com/percona-platform/dbaas-api/gen/controller"
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+	"gopkg.in/reform.v1"
+
+	dbaasv1beta1 "github.com/percona/pmm/api/managementpb/dbaas"
+	"github.com/percona/pmm/managed/models"
+	pmmversion "github.com/percona/pmm/version"
+)
+
+// Initializer initializes dbaas feature
+type Initializer struct {
+	db *reform.DB
+	l  *logrus.Entry
+
+	dbaasClient    dbaasClient
+	grafanaClient  grafanaClient
+	versionService versionService
+
+	enabled bool
+	cancel  func()
+	m       sync.Mutex
+}
+
+const defaultClusterName = "default-pmm-cluster"
+
+var errClusterExists = errors.New("cluster already exists")
+
+// NewInitializer returns initialized Initializer structure
+func NewInitializer(db *reform.DB, client dbaasClient, grafanaClient grafanaClient, versionService versionService) *Initializer {
+	l := logrus.WithField("component", "dbaas_initializer")
+	return &Initializer{
+		db:             db,
+		l:              l,
+		dbaasClient:    client,
+		grafanaClient:  grafanaClient,
+		versionService: versionService,
+	}
+}
+
+// Update updates current dbaas settings
+func (in *Initializer) Update(ctx context.Context) error {
+	settings, err := models.GetSettings(in.db)
+	if err != nil {
+		in.l.Errorf("Failed to get settings: %+v.", err)
+		return err
+	}
+	if settings.DBaaS.Enabled {
+		return in.Enable(ctx)
+	}
+	return in.Disable(ctx)
+}
+
+// Enable enables dbaas feature and connects to dbaas-controller
+func (in *Initializer) Enable(ctx context.Context) error {
+	in.m.Lock()
+	defer in.m.Unlock()
+	if in.enabled {
+		return nil
+	}
+	timeoutCtx, cancel := context.WithTimeout(ctx, 20*time.Second)
+	err := in.dbaasClient.Connect(timeoutCtx)
+	cancel()
+	if err != nil {
+		return err
+	}
+	ctx, in.cancel = context.WithCancel(ctx)
+
+	in.enabled = true
+	kubeConfig, err := in.dbaasClient.GetKubeConfig(ctx, &dbaascontrollerv1beta1.GetKubeconfigRequest{})
```

Comment:
can we extract everything below this line to a separate method?

Reply:
And should we do it each time?

Reply:
I moved it to a separate function. 

> And should we do it each time?

Yes, but we do not do anything else once PMM has registered cluster.